src/linear_algebra/free_module.lean

```diff
@@ -207,4 +206,17 @@ of_basis (basis.of_vector_space R M)
 
 end division_ring
 
+section integer
+
+variables [add_comm_group M] [module.finite ℤ M] [module.free ℤ M]
+variables [add_comm_group N] [module.finite ℤ N] [module.free ℤ N]
+
+instance : module.finite ℤ (M →+ N) :=
+module.finite.equiv add_monoid_hom_lequiv_int.symm
+
+instance : module.free ℤ (M →+ N) :=
+module.free.of_equiv add_monoid_hom_lequiv_int.symm
```

Comment:
Are these free over other rings too? [`add_monoid_hom.module`](https://leanprover-community.github.io/mathlib_docs/algebra/module/hom.html#add_monoid_hom.module) says that `M →+ N` is a module over many rings, and presumably it would be straightforward to adapt `add_monoid_hom_lequiv_int`

Reply:
No, I think it depends on `R` being finite/free as `ℤ` algebra/module (or something like that).

`add_monoid_hom_lequiv_int` can be probably generalized a little bit, to `Z/n` or `ℚ`, but the statement become less interesting and I don't think anybody will ever use them.

Reply:
> `add_monoid_hom_lequiv_int` can be probably generalized a little bit, to `Z/n` or ℚ`,

It generalizes to an arbitrary `R`, see #9233. Arguably the `ℤ` in that position muddled what the statement actually was.

Reply:
Ah, I see the issue now with my suggestion; this code will only work when `R = Z`:

https://github.com/leanprover-community/mathlib/blob/f536b4f057a51e4e6248c35ffce093c0351e5efd/src/linear_algebra/free_module.lean#L169-L174

So what you have is probably fine as is.

Reply:
Ah sure, you can generalize it that way. I thought you wanted to consder `A →ₗ[R] B`, with `A` an `R`-module too. `(A →ₗ[ℤ] B)` is surely finite over `R`, being finite over `ℤ` (this should be added in general if not already there). I don't know about freeness.

Reply:
I am almost sure it's false in general. Take for `R` the ring of integer of any number field of class number bigger than `1` (so `R` is a Dedekind domain but not a PID, for example `R = ℤ[i√5]`). Now, let `M` be a finitely generated projective but not free `R`-modules (they exist since the class group = picard group is not trivial). Let `N` be `M →ₗ[R] R`. Now, `M` is projective, so it is `R` isomorphic to its double `R` dual, that is `N →ₗ[R] R`: in particular `N →ₗ[R] R` it's not free as `R`-module.

Reply:
The following seems to give an explicit counterexample

Take `R = ℤ[i√5]`: it is the ring of integer of a number field of class number bigger than `1` (so `R` is a Dedekind domain but not a PID). Now, let `M` be a finitely generated projective but not free `R`-modules (they exist since the class group = picard group is not trivial). Now, let us consider `R →ₗ[ℤ] M` (pay attention that I am considering `ℤ`-linear maps here). Since `R` is free as a `ℤ`-module (of rank two), this is isomorphic to `M^2`as `ℤ`-module. As you noted, this isomorphism is automatically `R`-linear, but `M^2` is not free as `R`-module.